File: apps/rivethub-web/src/lib/ask-user.ts
/**
 * Extract ask-user prompts from ask-user tool shapes.
 * Supports:
 * - Claude AskUserQuestion: { questions: [{ question, header, multiSelect,
 *   options: [{ label, description }] }] }
 * - Grok ask_user_question: flat options/choices (strings or {label})
 * - RivetOS ask_user: { choices: string[] }
 * Missing/malformed args → empty array (never throws).
 */

import { normalizeToolName } from './tool-titles.js'

const ASK_TOOL_NAMES = new Set(['ask_user', 'ask_user_question', 'askuserquestion'])

export function isAskUserTool(name: string): boolean {
  const n = normalizeToolName(name).toLowerCase().replace(/\s+/g, '_')
  return ASK_TOOL_NAMES.has(n)
}

function labelFromOption(opt: unknown): string | undefined {
  if (typeof opt === 'string' && opt.trim()) return opt.trim()
  if (opt && typeof opt === 'object') {
    const o = opt as Record<string, unknown>
    if (typeof o.label === 'string' && o.label.trim()) return o.label.trim()
    if (typeof o.value === 'string' && o.value.trim()) return o.value.trim()
    if (typeof o.text === 'string' && o.text.trim()) return o.text.trim()
  }
  return undefined
}

// ---------------------------------------------------------------------------
// Structured extraction — the composer's ask card.

export interface AskOption {
  label: string
  description?: string
}

export interface AskQuestion {
  /** The question text itself (may be absent on bare option lists). */
  question?: string
  /** Claude's short chip label ("Auth method") — used as the answer prefix
   *  when multiple questions are answered at once. */
  header?: string
  multiSelect: boolean
  options: AskOption[]
}

function optionFrom(opt: unknown): AskOption | undefined {
  const label = labelFromOption(opt)
  if (!label) return undefined
  if (opt && typeof opt === 'object') {
    const d = (opt as Record<string, unknown>).description
    if (typeof d === 'string' && d.trim()) return { label, description: d.trim() }
  }
  return { label }
}

function optionsFromArray(arr: unknown): AskOption[] {
  if (!Array.isArray(arr)) return []
  const seen = new Set<string>()
  const out: AskOption[] = []
  for (const item of arr) {
    const o = optionFrom(item)
    if (!o || seen.has(o.label)) continue
    seen.add(o.label)
    out.push(o)
    if (out.length >= 10) break
  }
  return out
}

function questionFrom(q: unknown): AskQuestion | undefined {
  if (!q || typeof q !== 'object') return undefined
  const qo = q as Record<string, unknown>
  const options = [...optionsFromArray(qo.options), ...optionsFromArray(qo.choices)]
  if (options.length === 0) return undefined
  const question =
    typeof qo.question === 'string' && qo.question.trim() ? qo.question.trim() : undefined
  const header = typeof qo.header === 'string' && qo.header.trim() ? qo.header.trim() : undefined
  return { question, header, multiSelect: qo.multiSelect === true, options }
}

/**
 * Parse tool args into structured ask questions. Safe on any unknown input;
 * empty array when nothing extractable (caller falls back to nothing).
 */
export function extractAskUserQuestions(args: unknown): AskQuestion[] {
  if (args == null) return []
  let root: unknown = args
  if (typeof args === 'string') {
    try {
      root = JSON.parse(args) as unknown
    } catch {
      return []
    }
  }
  if (typeof root !== 'object' || root === null) return []
  const obj = root as Record<string, unknown>

  // Nested Claude shape — up to 4 questions per call.
  if (Array.isArray(obj.questions)) {
    const nested = obj.questions.map(questionFrom).filter((q): q is AskQuestion => q !== undefined)
    if (nested.length) return nested.slice(0, 4)
  }

  // Flat single-question shapes (grok / rivetos / yes_no).
  const flat = questionFrom(obj)
  if (flat) return [flat]
  if (obj.type === 'yes_no') {
    const question =
      typeof obj.question === 'string' && obj.question.trim() ? obj.question.trim() : undefined
    return [{ question, multiSelect: false, options: [{ label: 'Yes' }, { label: 'No' }] }]
  }
  return []
}

/**
 * From a live tool stack (newest last), the last ask-user tool's structured
 * questions — running or done (headless CLIs auto-complete the tool; the
 * answer is the next user turn).
 */
export function questionsFromLiveTools(
  tools: ReadonlyArray<{ name: string; args?: unknown; status: string }>,
): AskQuestion[] {
  for (let i = tools.length - 1; i >= 0; i--) {
    const t = tools[i]
    if (!isAskUserTool(t.name)) continue
    const qs = extractAskUserQuestions(t.args)
    if (qs.length) return qs
  }
  return []
}

File: apps/rivethub-android/app/src/main/java/io/rivethub/app/plane/AskUser.kt
package io.rivethub.app.plane

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.contentOrNull

data class AskOption(val label: String, val description: String? = null)

data class AskQuestion(
    val question: String? = null,
    val header: String? = null,
    val multiSelect: Boolean = false,
    val options: List<AskOption> = emptyList(),
)

data class AskUserCard(val questions: List<AskQuestion>)

data class LiveTool(val name: String, val args: JsonElement? = null, val status: String = "running")

private val ASK_JSON = Json { ignoreUnknownKeys = true; isLenient = true }

private val ASK_TOOL_NAMES = setOf("ask_user", "ask_user_question", "askuserquestion")

fun normalizeToolName(raw: String): String {
    val t = raw.trim().trimStart { !it.isLetter() && !it.isDigit() && it != '_' }.trim()
    return t.ifEmpty { raw.trim().ifEmpty { "tool" } }
}

fun isAskUserTool(name: String): Boolean {
    val n = normalizeToolName(name).lowercase().replace(Regex("\\s+"), "_")
    return n in ASK_TOOL_NAMES
}

private fun labelFromOption(opt: JsonElement): String? {
    if (opt is JsonPrimitive && opt.contentOrNull?.trim()?.isNotEmpty() == true) return opt.content.trim()
    if (opt is JsonObject) {
        val label = opt.str("label") ?: opt.str("value") ?: opt.str("text")
        if (label != null) return label
    }
    return null
}

private fun optionFrom(opt: JsonElement): AskOption? {
    val label = labelFromOption(opt) ?: return null
    if (opt is JsonObject) {
        val d = opt.str("description")
        if (d != null) return AskOption(label, d)
    }
    return AskOption(label)
}

private fun optionsFromArray(arr: JsonElement?): List<AskOption> {
    if (arr !is JsonArray) return emptyList()
    val seen = HashSet<String>()
    val out = ArrayList<AskOption>()
    for (item in arr) {
        val o = optionFrom(item) ?: continue
        if (!seen.add(o.label)) continue
        out += o
        if (out.size >= 20) break
    }
    return out
}

private fun questionFrom(q: JsonElement): AskQuestion? {
    if (q !is JsonObject) return null
    val options = optionsFromArray(q["options"]) + optionsFromArray(q["choices"])
    if (options.isEmpty()) return null
    return AskQuestion(
        question = q.str("question"),
        header = q.str("header"),
        multiSelect = (q["multiSelect"] as? JsonPrimitive)?.booleanOrNull == true,
        options = options,
    )
}

/** Parse tool args into structured ask questions. Empty on anything unusable. */
fun extractAskUserQuestions(args: Any?): List<AskQuestion> {
    val root = toJson(args) ?: return emptyList()
    if (root !is JsonObject) return emptyList()
    val nested = root["questions"]
    if (nested is JsonArray) {
        val qs = nested.mapNotNull { questionFrom(it) }
        if (qs.isNotEmpty()) return qs.take(4)
    }
    questionFrom(root)?.let { return listOf(it) }
    if (root.str("type") == "yes_no") {
        return listOf(
            AskQuestion(
                question = root.str("question"),
                multiSelect = false,
                options = listOf(AskOption("Yes"), AskOption("No")),
            ),
        )
    }
    return emptyList()
}

fun questionsFromLiveTools(tools: List<LiveTool>): List<AskQuestion> {
    for (i in tools.indices.reversed()) {
        val t = tools[i]
        if (!isAskUserTool(t.name)) continue
        val qs = extractAskUserQuestions(t.args)
        if (qs.isNotEmpty()) return qs
    }
    return emptyList()
}

fun cardFromLiveTools(tools: List<LiveTool>): AskUserCard? {
    val qs = questionsFromLiveTools(tools)
    return if (qs.isEmpty()) null else AskUserCard(qs)
}

/**
 * Compose the outgoing answer from option picks (keyed by question index)
 * and free text. Multi-question picks are prefixed so the agent can match
 * them up. Empty when there is nothing to send.
 */
fun composeAskAnswer(
    questions: List<AskQuestion>,
    picked: Map<Int, List<String>>,
    own: String,
): String {
    val answered = questions.mapIndexed { qi, q -> q to (picked[qi] ?: emptyList()) }
        .filter { it.second.isNotEmpty() }
    val parts = ArrayList<String>()
    if (answered.isNotEmpty()) {
        parts += if (questions.size == 1) {
            answered[0].second.joinToString(", ")
        } else {
            answered.joinToString("\n") { (q, labels) ->
                val prefix = q.header ?: q.question
                if (prefix != null) "$prefix: ${labels.joinToString(", ")}" else labels.joinToString(", ")
            }
        }
    }
    val free = own.trim()
    if (free.isNotEmpty()) parts += free
    return parts.joinToString("\n")
}

private fun toJson(args: Any?): JsonElement? {
    if (args == null) return null
    if (args is JsonElement) return args
    if (args is String) {
        return runCatching { ASK_JSON.parseToJsonElement(args) }.getOrNull()
    }
    return null
}

private fun JsonObject.str(key: String): String? =
    (this[key] as? JsonPrimitive)?.contentOrNull?.trim()?.takeIf { it.isNotEmpty() }
